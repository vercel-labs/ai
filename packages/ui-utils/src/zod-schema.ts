import { JSONSchema7 } from 'json-schema';
import { z } from 'zod';
import zodToJsonSchema from 'zod-to-json-schema';
import { jsonSchema, Schema } from './schema';

export function zodSchema<OBJECT>(
  zodSchema: z.Schema<OBJECT, z.ZodTypeDef, any>,
): Schema<OBJECT> {
  return jsonSchema(
    zodToJsonSchema(zodSchema, {
      $refStrategy: 'none', // no references (to support openapi conversion for google)
      target: 'jsonSchema7', // note: openai mode breaks various gemini conversions
    }) as JSONSchema7,
    {
      validate: value => {
        const result = zodSchema.safeParse(value);
        return result.success
          ? { success: true, value: result.data }
          : { success: false, error: result.error };
      },
    },
  );
}
