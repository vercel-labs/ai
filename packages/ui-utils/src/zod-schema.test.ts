import { z } from 'zod';
import { zodSchema } from './zod-schema';
import { safeParseJSON } from '@ai-sdk/provider-utils';

describe('zodSchema', () => {
  describe('json schema conversion', () => {
    it('should create a schema with simple types', () => {
      const schema = zodSchema(
        z.object({
          text: z.string(),
          number: z.number(),
        }),
      );

      expect(schema.jsonSchema).toMatchSnapshot();
    });

    it('should support optional fields in object', () => {
      const schema = zodSchema(
        z.object({
          required: z.string(),
          optional: z.string().optional(),
        }),
      );

      expect(schema.jsonSchema).toMatchSnapshot();
    });

    it('should support optional fields with descriptions in object', () => {
      const schema = zodSchema(
        z.object({
          required: z.string().describe('Required description'),
          optional: z.string().optional().describe('Optional description'),
        }),
      );

      expect(schema.jsonSchema).toMatchSnapshot();
    });

    it('should support arrays', () => {
      const schema = zodSchema(
        z.object({
          items: z.array(z.string()),
        }),
      );

      expect(schema.jsonSchema).toMatchSnapshot();
    });

    it('should support optional arrays', () => {
      const schema = zodSchema(
        z.object({
          items: z.array(z.string()).optional(),
        }),
      );

      expect(schema.jsonSchema).toMatchSnapshot();
    });

    it('should support required enums', () => {
      const schema = zodSchema(
        z.object({
          type: z.enum(['a', 'b', 'c']),
        }),
      );

      expect(schema.jsonSchema).toMatchSnapshot();
    });

    it('should support optional enums', () => {
      const schema = zodSchema(
        z.object({
          type: z.enum(['a', 'b', 'c']).optional(),
        }),
      );

      expect(schema.jsonSchema).toMatchSnapshot();
    });

    it('should duplicate referenced schemas (and not use references)', () => {
      const Inner = z.object({
        text: z.string(),
        number: z.number(),
      });

      const schema = zodSchema(
        z.object({
          group1: z.array(Inner),
          group2: z.array(Inner),
        }),
      );

      expect(schema.jsonSchema).toMatchSnapshot();
    });
  });

  describe('output validation', () => {
    it('should validate output with transform', () => {
      const schema = zodSchema(
        z.object({
          user: z.object({
            id: z.string().transform(val => parseInt(val, 10)),
            name: z.string(),
          }),
        }),
      );

      const result = safeParseJSON({
        text: '{"user": {"id": "123", "name": "John"}}',
        schema,
      });

      expect(result).toStrictEqual({
        success: true,
        value: { user: { id: 123, name: 'John' } },
        rawValue: { user: { id: '123', name: 'John' } },
      });
    });
  });
});
