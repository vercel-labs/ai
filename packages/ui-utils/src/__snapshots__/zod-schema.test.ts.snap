// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`zodSchema > json schema conversion > should create a schema with simple types 1`] = `
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "additionalProperties": false,
  "properties": {
    "number": {
      "type": "number",
    },
    "text": {
      "type": "string",
    },
  },
  "required": [
    "text",
    "number",
  ],
  "type": "object",
}
`;

exports[`zodSchema > json schema conversion > should duplicate referenced schemas (and not use references) 1`] = `
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "additionalProperties": false,
  "properties": {
    "group1": {
      "items": {
        "additionalProperties": false,
        "properties": {
          "number": {
            "type": "number",
          },
          "text": {
            "type": "string",
          },
        },
        "required": [
          "text",
          "number",
        ],
        "type": "object",
      },
      "type": "array",
    },
    "group2": {
      "items": {
        "additionalProperties": false,
        "properties": {
          "number": {
            "type": "number",
          },
          "text": {
            "type": "string",
          },
        },
        "required": [
          "text",
          "number",
        ],
        "type": "object",
      },
      "type": "array",
    },
  },
  "required": [
    "group1",
    "group2",
  ],
  "type": "object",
}
`;

exports[`zodSchema > json schema conversion > should support arrays 1`] = `
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "additionalProperties": false,
  "properties": {
    "items": {
      "items": {
        "type": "string",
      },
      "type": "array",
    },
  },
  "required": [
    "items",
  ],
  "type": "object",
}
`;

exports[`zodSchema > json schema conversion > should support optional arrays 1`] = `
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "additionalProperties": false,
  "properties": {
    "items": {
      "items": {
        "type": "string",
      },
      "type": "array",
    },
  },
  "type": "object",
}
`;

exports[`zodSchema > json schema conversion > should support optional enums 1`] = `
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "additionalProperties": false,
  "properties": {
    "type": {
      "enum": [
        "a",
        "b",
        "c",
      ],
      "type": "string",
    },
  },
  "type": "object",
}
`;

exports[`zodSchema > json schema conversion > should support optional fields in object 1`] = `
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "additionalProperties": false,
  "properties": {
    "optional": {
      "type": "string",
    },
    "required": {
      "type": "string",
    },
  },
  "required": [
    "required",
  ],
  "type": "object",
}
`;

exports[`zodSchema > json schema conversion > should support optional fields with descriptions in object 1`] = `
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "additionalProperties": false,
  "properties": {
    "optional": {
      "description": "Optional description",
      "type": "string",
    },
    "required": {
      "description": "Required description",
      "type": "string",
    },
  },
  "required": [
    "required",
  ],
  "type": "object",
}
`;

exports[`zodSchema > json schema conversion > should support required enums 1`] = `
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "additionalProperties": false,
  "properties": {
    "type": {
      "enum": [
        "a",
        "b",
        "c",
      ],
      "type": "string",
    },
  },
  "required": [
    "type",
  ],
  "type": "object",
}
`;
